reporter/benchmark.go

```diff
@@ -0,0 +1,255 @@
+	Symbol  string
+}
+
+func (r *BenchmarkReporter) ReportFallbackSymbol(frameID libpf.FrameID, symbol string) {
+	r.store("FallbackSymbol", &fallbackSymbol{
+		FrameID: frameID,
+		Symbol:  symbol,
+	})
+	r.rep.ReportFallbackSymbol(frameID, symbol)
+}
+
+func (r *BenchmarkReporter) ExecutableMetadata(ctx context.Context, fileID libpf.FileID,
+	fileName, buildID string, interp libpf.InterpreterType, open ExecutableOpener) {
+	r.rep.ExecutableMetadata(ctx, fileID, fileName, buildID, interp, open)
+}
+
+type frameMetadata struct {
+	FileID         libpf.FileID
+	AddressOrLine  libpf.AddressOrLineno
+	LineNumber     util.SourceLineno
+	FunctionOffset uint32
+	FunctionName   string
+	FilePath       string
+}
+
+func (r *BenchmarkReporter) FrameMetadata(fileID libpf.FileID, addressOrLine libpf.AddressOrLineno,
+	lineNumber util.SourceLineno, functionOffset uint32, functionName, filePath string) {
+	r.store("FrameMetadata", &frameMetadata{
+		FileID:         fileID,
+		AddressOrLine:  addressOrLine,
+		LineNumber:     lineNumber,
+		FunctionOffset: functionOffset,
+		FunctionName:   functionName,
+		FilePath:       filePath,
+	})
+	r.rep.FrameMetadata(fileID, addressOrLine, lineNumber, functionOffset, functionName, filePath)
+}
+
+func (r *BenchmarkReporter) ReportHostMetadata(metadata map[string]string) {
+	r.rep.ReportHostMetadata(metadata)
+}
+
+func (r *BenchmarkReporter) ReportHostMetadataBlocking(ctx context.Context,
+	metadataMap map[string]string, maxRetries int, waitRetry time.Duration) error {
+	return r.rep.ReportHostMetadataBlocking(ctx, metadataMap, maxRetries, waitRetry)
+}
+
+type metrics struct {
+	Timestamp uint32
+	IDs       []uint32
+	Values    []int64
+}
+
+func (r *BenchmarkReporter) ReportMetrics(timestamp uint32, ids []uint32, values []int64) {
+	r.store("Metrics", &metrics{
+		Timestamp: timestamp,
+		IDs:       ids,
+		Values:    values,
+	})
+	r.rep.ReportMetrics(timestamp, ids, values)
+}
+
+func (r *BenchmarkReporter) Stop() {
+	r.rep.Stop()
+}
+
+func (r *BenchmarkReporter) GetMetrics() Metrics {
+	return r.rep.GetMetrics()
+}
+
+func NewBenchmarkReporter(benchDataDir string, rep Reporter) (*BenchmarkReporter, error) {
+	r := &BenchmarkReporter{
+		benchDataDir: benchDataDir,
+		rep:          rep,
+	}
+	r.uid, r.gid = originUser()
+
+	if err := os.MkdirAll(benchDataDir, 0o755); err != nil {
+		return nil, err
+	}
+
+	if r.uid != 0 || r.gid != 0 {
+		changeDirOwner(benchDataDir, r.uid, r.gid)
+	}
+
+	// Just for storing the initial timestamp.
+	r.store("Start", libpf.Void{})
+
+	return r, nil
+}
+
+func originUser() (uid, gid int) {
+	if uidStr := os.Getenv("SUDO_UID"); uidStr != "" {
+		uid, _ = strconv.Atoi(uidStr)
+	}
+	if gidStr := os.Getenv("SUDO_GID"); gidStr != "" {
+		gid, _ = strconv.Atoi(gidStr)
+	}
+	return
+}
+
+type traceEvent struct {
+	Trace *libpf.Trace
+	Meta  *TraceEventMeta
+}
+
+var counter atomic.Uint64
+
+// store stores data as JSON.
+func (r *BenchmarkReporter) store(name string, data any) {
```

Comment:
What's the rationale for creating one file per message? Intuitively I would've probably gone with a single combined ndjson file. That'd be simpler and also avoid the need for reading all files into memory in the replay reporter to sort them by timestamp: you can simply read it as a stream.

Reply:
:+1: Makes sense. I basically re-used the code / strategy from the protobuf message storing without thinking much about it.

Reply:
This is done.